### PR TITLE
[gc] Register a critical region when executing managed allocators.

### PR DESCRIPTION
**Pull request by alexrp**
Previously, if we stopped a thread and it had a stack looking like this:

    ...
    profiler_signal_handler
    <signal handler called>
    managed_allocator
    ...

We would fail to identify the fact that we've just interrupted a thread that's
in a managed allocator (uninterruptible code) because we only look at the very
latest instruction pointer of the thread, which will be pointing into
profiler_signal_handler or some other function called by it. So we would
happily continue along with the STW process and proceed to doing the actual GC,
where we would see a broken heap.

We could solve this by unwinding the stack and checking all frames, but that's
complicated and error-prone. Instead, register a critical region while the
managed allocator runs. This way, if we don't identify it by instruction pointer, we
will identify it by the fact that the thread is in a critical region.

### CODE REVIEW

**Comment by alexrp** [comment]
@kumpera @schani review?

Note that I've only enabled the critical region variant of the managed allocator in the sample profiling case. I don't know whether the slowdown is too much to warrant enabling it in all cases (which would be nice since our profiler is not necessarily the only thing that could be bombarding us with signals).

**Comment by alexrp** [comment]
I used this test: https://gist.github.com/alexrp/b71138090ae4b2e880d14cb42cbf7239

Ran `time mono --profile=log:sample=cycles/10000,nocalls,noalloc gc.exe`.

Before:

    real    0m0.625s
    user    0m0.652s
    sys     0m0.032s

After:

    real    0m0.883s
    user    0m0.948s
    sys     0m0.012s

Without managed allocators (`MONO_GC_DEBUG=no-managed-allocator`):

    real    0m7.678s
    user    0m8.529s
    sys     0m0.024s

**Comment by alexrp** [comment]
builddeb

**Comment by schani** [comment]
Looks good.

**Comment by alexrp** [comment]
builddeb

**Comment by alexrp** [comment]
builddeb

File: mono/metadata/sgen-mono.c [hunk 1]
```diff
@@ -987,19 +987,29 @@ mono_gc_free_fixed (void* addr)
 
 static MonoMethod* alloc_method_cache [ATYPE_NUM];
 static MonoMethod* slowpath_alloc_method_cache [ATYPE_NUM];
+static MonoMethod* critical_alloc_method_cache [ATYPE_NUM];
```
Comment (kumpera) [inline mono/metadata/sgen-mono.c]: Nothing assigns to critical_alloc_method_cache

**Comment by alexrp** [comment]
@kumpera @schani rebased, can you have another look?

**Comment by schani** [comment]
If it's just a rebase then I'm still ok with it.  Like before, I'd like to see benchmarking results, but we're still blocked on that.

File: mono/metadata/sgen-mono.c [hunk 2]
```diff
@@ -1116,6 +1142,16 @@ create_allocator (int atype, gboolean slowpath)
...
 	}
 
+	EMIT_TLS_ACCESS_VAR (mb, thread_var);
+
+#ifdef MANAGED_ALLOCATOR_CAN_USE_CRITICAL_REGION
+	EMIT_TLS_ACCESS_IN_CRITICAL_REGION_ADDR (mb, thread_var);
+	mono_mb_emit_byte (mb, CEE_LDC_I4_1);
+	mono_mb_emit_byte (mb, MONO_CUSTOM_PREFIX);
+	mono_mb_emit_byte (mb, CEE_MONO_ATOMIC_STORE_I4);
+	mono_mb_emit_i4 (mb, MONO_MEMORY_BARRIER_NONE);
```
Comment (alexrp) [inline mono/metadata/sgen-mono.c]: @schani to clarify, the only functional change is here (was an acquire barrier before).

**Comment by kumpera** [comment]
LGTM.
If perf numbers are good, lets merge this puppy.

**Comment by kumpera** [comment]
builddeb

**Comment by lewurm** [comment]
FYI, `builddeb` command doesn't work right now, as it isn't migrated yet to the new jenkins instance /cc @directhex

**Comment by alexrp** [comment]
So, do we merge this now or wait until the benchmarking bots are back up? I suppose if we merge right away, we can just get perf results when the bots are back up by posting a PR with a revert of these changes.

**Comment by lewurm** [comment]
@alexrp: I would prefer to wait, as it could mask other perf issues. The service should be back hopefully by the end of this week (*fingers crossed*).  OTOH, it's fixing a bug, so if you need it to e.g. make the profiler less crashy, please merge it :)

**Comment by akoeplinger** [comment]
builddeb

**Comment by alexrp** [comment]
builddeb

**Comment by akoeplinger** [comment]
@alexrp looks like you need to rebase to get some packaging fixes from master.

**Comment by alexrp** [comment]
builddeb

**Comment by alexrp** [comment]
builddeb

**Comment by alexrp** [comment]
@schani @lewurm benchmark results are up if you guys want to have a look.

**Comment by alexrp** [comment]
OK, I split the removal of the pointless memory barrier into #3083.

**Comment by alexrp** [comment]
builddeb

**Comment by kumpera** [comment]
perf numbers looked good from the previous run. can we merge this?

**Comment by lewurm** [comment]
builddeb

**Comment by lewurm** [comment]
performance bot failed to post update:
linux/amd64 sgen-serial: http://open.xamarin.com/benchmarker/front-end/pullrequest.html#id=337
linux/amd64 sgen-concurrent: http://open.xamarin.com/benchmarker/front-end/pullrequest.html#id=334
linux/arm32 sgen-serial: http://open.xamarin.com/benchmarker/front-end/pullrequest.html#id=329

**Comment by lewurm** [comment]
@monojenkins merge

**Pull request merged**